### PR TITLE
Feature/custom mesh primitive attrs

### PR DESCRIPTION
**Pull request by FreakTheMighty**
## Overview
This change allows custom attributes to be parsed and stored in a dictionary object.

This pull request changes the glTF `Attributes` from an custom object, `GltfMeshPrimitiveAttributes`  to a dictionary to support arbitrary attributes. `GltfMeshPrimitiveAttributes` is now used as a look up for standard attribute keys.

## Changes
- Fixes: #6953 

## Verification

Two playmode tests have been added. `TestGltfLoads` and `TestGltfCustomAttributes` to provide some basic coverage for the existing glTF loading process and the new addition with custom attributes.

### CODE REVIEW

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/GltfTests.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+using System.Collections;
+using System.IO;
+using System.Threading.Tasks;
+using UnityEngine.TestTools;
+
+namespace Microsoft.MixedReality.Toolkit.Tests
+{
+    public class GltfTests
+    {
+        const string relativePath = "Assets/MixedRealityToolkit.Examples/Demos/Gltf/Models/Avocado/glTF/AvocadoCustomAttr.gltf";
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/GltfTests.cs]: To make the tests more robust to file and folder name changes, would it be possible to load this path a different way, perhaps by GUID like in https://github.com/microsoft/MixedRealityToolkit-Unity/pull/7441?

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/GltfTests.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+        private IEnumerator WaitForTask(Task task)
+        {
+            while (!task.IsCompleted) { yield return null; }
+            if (task.IsFaulted) { throw task.Exception; }
+            yield return null;
+        }
+
+        [UnityTest]
+        public IEnumerator TestGltfLoads()
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/GltfTests.cs]: Please add summaries to these tests, so it's clear what they're testing

File: Assets/MixedRealityToolkit/Utilities/Gltf/Serialization/GltfUtility.cs [hunk 3]
```diff
@@ -394,5 +394,62 @@ private static string GetGltfNodeName(string jsonString)
...
+
+                reformatted += $"{{\"key\":\"{key}\", \"value\":{value}}},";
+            }
+            reformatted = reformatted.TrimEnd(',');
+            reformatted += "]}";
+            return reformatted;
+        }
+
+        [System.Serializable]
+        public class StringKeyValue
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Utilities/Gltf/Serialization/GltfUtility.cs]: Does this need to be `public`? I couldn't find a reference to it outside this class.
If it does, please [add summaries](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/CodingGuidelines.html#function--method-summary-headers) to the class and properties.

File: Assets/MixedRealityToolkit/Utilities/Gltf/Serialization/GltfUtility.cs [hunk 4]
```diff
@@ -394,5 +394,62 @@ private static string GetGltfNodeName(string jsonString)
...
+        {
+            public string key;
+            public int value;
+        }
+
+        [System.Serializable]
+        private class StringIntKeyValueArray
+        {
+            public StringKeyValue[] items = new StringKeyValue[0];
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Utilities/Gltf/Serialization/GltfUtility.cs]: nit: consider [Array.Empty](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1825) instead of a zero-length array allocation

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/GltfTests.cs (hunk 2 above)
Comment (FreakTheMighty) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/GltfTests.cs]: Done!

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/GltfTests.cs (hunk 1 above)
Comment (FreakTheMighty) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/GltfTests.cs]: Sounds good

**Comment by FreakTheMighty** [comment]
@keveleigh thank's for the notes. I've made the changes you suggested. Not sure why the build is failing though.

**Comment by keveleigh** [comment]
@FreakTheMighty thanks! I'll kick off another build and see what's up

**Comment by FreakTheMighty** [comment]
@keveleigh seems like the build is still stalling. Let me dig in a little and see what could be causing it.

**Comment by FreakTheMighty** [comment]
@keveleigh does anything pop out to you? I see lots of these errors

```
Failed to start keyword recognizer. Are microphone permissions granted? Exception: UnityEngine.UnityException: Speech recognition is not supported on this machine.
```

Those errors seem to be present in successful runs as well. Should I throw a `yield return null` in before the asserts?

**Comment by keveleigh** [comment]
@FreakTheMighty Nothing is jumping out at me initially...we do have [a powershell script](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/mrtk_development/scripts/test/run_playmode_tests.ps1) that runs the tests the same way we run them in validation. sometimes play mode tests act differently depending on how they're run. Taking a look locally as well

**Comment by FreakTheMighty** [comment]
After clearing some project settings I started to see the stall when running in the editor. I added a yield statement and the test appears to complete.

Do you mind kicking of the build again?

**Comment by FreakTheMighty** [comment]
@keveleigh that seems to have done the trick! 🤷‍♂

**Comment by keveleigh** [comment]
@FreakTheMighty Nice job figuring that out! Not sure exactly why it was needed, but, as you said: 🤷‍♂️

**Comment by FreakTheMighty** [comment]
@keveleigh what’s the next step to get this merged?

File: Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs [hunk 5]
```diff
@@ -9,21 +9,17 @@ namespace Microsoft.MixedReality.Toolkit.Utilities.Gltf.Schema
     /// Common mesh primitive attributes.
     /// https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/schema/mesh.primitive.schema.json
     /// </summary>
-    /// <remarks>
-    /// Application specific semantics are not supported
-    /// </remarks>
-    [Serializable]
     public class GltfMeshPrimitiveAttributes
     {
-        public int POSITION = -1;
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs]: Removing these public fields is a [breaking change](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/BreakingChanges.html). I do think this is okay in this case though, since this feels largely like an internal implementation detail for the glTF loader. Interested to get other opinions though.

File: Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitive.cs [hunk 6]
```diff
@@ -53,7 +53,7 @@ public class GltfMeshPrimitive : GltfProperty
         /// A dictionary object, where each key corresponds to mesh attribute semantic
         /// and each value is the index of the accessor containing attribute's data.
         /// </summary>
-        public GltfMeshPrimitiveAttributes Attributes { get; internal set; }
+        public Dictionary<string, int> Attributes { get; internal set; }
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitive.cs]: Same comment as https://github.com/microsoft/MixedRealityToolkit-Unity/pull/7490#discussion_r388562897.

Comment (keveleigh) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitive.cs]: This one feels like it might be easier to deprecate the existing property and create a new one than the other comment, though we wouldn't be able to reuse the name.

**Comment by keveleigh** [comment]
>what’s the next step to get this merged?

I want to get a few other eyes on this one, so I'll ping some people to review

File: Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitive.cs (hunk 6 above)
Comment (FreakTheMighty) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitive.cs]: What would the deprecated property do though? Wouldn’t it be fully broken? Also, what would we name the new property?

Comment (FreakTheMighty) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitive.cs]: What would the deprecated property do though? Wouldn’t it be fully broken? Also, what would we name the new property?

File: Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs (hunk 5 above)
Comment (FreakTheMighty) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs]: I could re-think things a little in an effort to keep it backwards compatible. Perhaps I could change these to properties that access an internal dictionary. . .

Just saying it out loud doesn’t sound quite right. This change certainly could bite someone, but it feel like a reasonable change to make with an appropriate release.

File: Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitive.cs (hunk 6 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitive.cs]: Hmm...possibly, if we keep filling both properties in the short term, it'd allow people to migrate if needed? Basically, keep https://github.com/microsoft/MixedRealityToolkit-Unity/pull/7490/files#diff-05951bac84393068d2888d4927ca4b57L202 around, but deprecate this property?
@davidkline-ms, what do you think? Might be more trouble than it's worth perf-wise to deserialize the glTF basically twice to keep this property around.

File: Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs (hunk 5 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs]: Yeah, agreed. I haven't fully dug in yet, but some of these possibly feel like classes that should have been `internal` to begin with. Not sure how useful this class would be outside the code that's loading the model

File: Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitive.cs (hunk 6 above)
Comment (FreakTheMighty) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitive.cs]: We could change this to an attribute, and return the ‘ GltfMeshPrimitiveAttributes’ instance hydrated with data from the dictionary.

File: Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs (hunk 5 above)
Comment (FreakTheMighty) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs]: Well, custom attributes is one use case that motivates making some of this public. I’m sure more thought could be put into exactly what the public interface should be.

Comment (FreakTheMighty) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs]: Well, custom attributes is one use case that motivates making some of this public. I’m sure more thought could be put into exactly what the public interface should be.

Comment (keveleigh) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs]: Oh good call, yeah

File: Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitive.cs (hunk 6 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitive.cs]: @keveleigh, is it possible to maintain both without deserializing twice by copying the objects?

Comment (keveleigh) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitive.cs]: @davidkline-ms yeah, it looks like @FreakTheMighty had a good idea on that above

**Comment by FreakTheMighty** [comment]
Thanks for all the feedback. I’m on the road for a few days, so it’s gonna take me a little while to incorporate the latest round of notes. When I get back to it though, I’ll make the small changes, then look into preventing breaking changes.

**Comment by FreakTheMighty** [comment]
@keveleigh @davidkline-ms I've made another pass that should be backwards compatible. Let me know what you think. The `GltfMeshPrimitiveAttributes` is now a subclass of a dictionary with glTF specific attributes accessible with the same names and types as before. I don't think we need any notes in `Updating.md` any more.

One thing I'm not totally sure about is that I've made the attributes read-only, but one could still set values via the dictionary. As I see it, I have three options:

1. Keep it as is 🤷‍♂ 
1. Implement the setter
1. Use a `ReadOnlyDictionary`

I think I lean towards the last option. It would be a _small_ breaking change, but I can't think of a reason that these should be settable.

File: Assets/MixedRealityToolkit/Utilities/Gltf/Serialization/ConstructGltf.cs [hunk 7]
```diff
@@ -246,6 +246,9 @@ private static async Task<Material> CreateMRTKShaderMaterial(GltfObject gltfObje
                 material.mainTexture = gltfObject.images[gltfMaterial.pbrMetallicRoughness.baseColorTexture.index].Texture;
             }
 
+            material.EnableKeyword("_DIRECTIONAL_LIGHT");
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Utilities/Gltf/Serialization/ConstructGltf.cs]: Is this related to this change?

File: Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs [hunk 8]
```diff
@@ -2,28 +2,114 @@
...
     /// <summary>
     /// Common mesh primitive attributes.
     /// https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/schema/mesh.primitive.schema.json
     /// </summary>
-    /// <remarks>
-    /// Application specific semantics are not supported
-    /// </remarks>
-    [Serializable]
-    public class GltfMeshPrimitiveAttributes
+    public class GltfMeshPrimitiveAttributes : Dictionary<string, int>
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs]: Nice change! From your other question, I think I'm good with this changing to a `ReadOnlyDictionary`. We can add a note to the changelist about it, but I'm not sure it makes sense for these to need updating after deserialization.

**Comment by keveleigh** [comment]
Also looks like some merge conflicts. I believe these are just from folder name changes, so rebasing or merging should resolve them cleanly.

File: Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs (hunk 8 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Schema/GltfMeshPrimitiveAttributes.cs]: Also, if the worry is people messing with the dictionary when it's returned, could this instead be implemented with a private dictionary in the class instead of the class being a dictionary itself? The constructor could still take one in and save it locally, and `TryGetDefault`'s `TryGetValue` would be called on the private dictionary instead of itself.

File: Assets/MixedRealityToolkit/Utilities/Gltf/Serialization/ConstructGltf.cs (hunk 7 above)
Comment (FreakTheMighty) [reply Assets/MixedRealityToolkit/Utilities/Gltf/Serialization/ConstructGltf.cs]: Doh, good catch. That should not be there.

**Comment by FreakTheMighty** [comment]
@keveleigh I've messed up the rebase, it looks like I've got some of the now removed directories still present in my branch. If you don't mind, I'm just gonna start clean and open a new pull request.

**Comment by keveleigh** [comment]
@FreakTheMighty no worries, go for it!

**Comment by FreakTheMighty** [comment]
closing in favor of #7535
